### PR TITLE
Pass Support for Inst Synthesis

### PR DESCRIPTION
**Pull request by zhengyangl**
This is just a working version (passes all the test cases), more test cases will be added tomorrow.

### CODE REVIEW

File: test/Pass/syn-inst-add-ext-use.ll [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+; RUN: llvm-as -o %t %s
+; RUN: opt -load %pass -souper %solver -souper-infer-inst -souper-synthesis-comps=add,const -S -o - %s | FileCheck %s
+
+; Investigation and improvement are needed. Current instruction synthesis can
+; not optimization instruction with outside uses smartly. This case shows this
+; issue: %a, %b and %c all have external uses, and they should not be swiped
+; out, thus synthesis a new instruction "add %x, 4" only for replacing %d is not
+; necessary. Now souper does this redundant optimization. 
+
```
Comment (zhengyangl) [inline test/Pass/syn-inst-add-ext-use.ll]: This describes the issue for not able to handle insts with external uses.

File: test/Pass/syn-inst-add.ll [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+; RUN: llvm-as -o %t %s
+; RUN: opt -load %pass -souper %solver -souper-infer-inst -souper-synthesis-comps=add,const -S -o - %s | FileCheck %s
+
+
+define i32 @foo(i32 %x) {
+entry:
+  %a = add i32 %x, 1
+  %b = add i32 %a, 1
+  %c = add i32 %b, 1
+  ;CHECK: add i32 %x, 4
```
Comment (regehr) [inline test/Pass/syn-inst-add.ll]: maybe also add a CHECK-NOT for each of the add-1 lines

File: lib/Pass/Pass.cpp [hunk 3]
```diff
@@ -279,12 +343,14 @@ struct SouperPass : public ModulePass {
       }
       if (!Cand.Mapping.RHS)
         continue;
-      // TODO: add non-const instruction support
+
       Instruction *I = Cand.Origin.getInstruction();
       Instruction *ReplacedInst = Cand.Origin.getInstruction();
+      IRBuilder<> Builder(ReplacedInst->getParent());
+      Builder.SetInsertPoint(ReplacedInst);
```
Comment (jeroenk) [inline lib/Pass/Pass.cpp]: Why not use ```IRBuilder<> Builder(ReplacedInst);```?

Comment (jeroenk) [reply lib/Pass/Pass.cpp]: Also. Why is the builder created here and not in the getValue function?

Comment (zhengyangl) [reply lib/Pass/Pass.cpp]: Thanks for this review!

Comment (zhengyangl) [reply lib/Pass/Pass.cpp]: Reason for IR builder created outside getValue() function: getValue() is a recursive function, we don't want to create an IR builder in each recursive call.

**Comment by zhengyangl** [comment]
Currently the annotation in these test cases does not correctly handle the checking on commutative operators (say mul i64 %x, 24). Need to find a solution for this issue.

**Comment by zhengyangl** [comment]
When dealing with binop instruction generation in getValue(), we can do a little bit canonicalization (Swap constants to the second operands if the operator is commutative).

**Comment by regehr** [comment]
Great. Yes, we should emit canonical code whenever possible!

**Comment by regehr** [comment]
LGTM, thanks!

**Pull request merged**